adminSiteClient/VariablesAnnotationPage.tsx

```diff
@@ -0,0 +1,927 @@
+
+            if (currentData !== undefined) {
+                runInAction(() => {
+                    console.log("Data fetched", currentData)
+                    this.resetViewStateAfterFetch()
+                    this.currentPagingOffset = this.desiredPagingOffset
+                    this.richDataRows = currentData.variables.map(
+                        parseVariableAnnotationsRow
+                    )
+                    this.numTotalRows = currentData.numTotalRows
+                })
+            }
+        })
+        // Add the disposer to the list of things we need to clean up on unmount
+        this.disposers.push(disposer)
+
+        this.getFieldDefinitions()
+    }
+
+    @action.bound private resetViewStateAfterFetch(): void {
+        this.undoSteps = []
+        this.redoSteps = []
+        this.selectedRow = undefined
+        this.grapherElement = undefined
+    }
+    @action.bound private loadGrapherJson(json: any): void {
+        const newConfig: GrapherProgrammaticInterface = {
+            ...json,
+            bounds: new Bounds(0, 0, 500, 400),
+            getGrapherInstance: (grapher: Grapher) => {
+                this.grapher = grapher
+            },
+        }
+        if (this.grapherElement) {
+            this.grapher.setAuthoredVersion(newConfig)
+            this.grapher.reset()
+            this.grapher.updateFromObject(newConfig)
+            this.grapher.downloadData()
+        } else this.grapherElement = <Grapher {...newConfig} />
+    }
+
+    @action private updatePreviewToRow(row: number): void {
+        const { richDataRows } = this
+        console.log("updating preview to row", row)
+        if (richDataRows === undefined) return
+
+        // Get the grapherConfig of the currently selected row and then
+        // merge it with the necessary partial information (e.g. variableId field)
+        // to get a config that actually works in all cases
+        const config = richDataRows[row].grapherConfig
+        const finalConfigLayer = getFinalConfigLayerForVariable(
+            richDataRows[row].id
+        )
+        const mergedConfig = merge(config, finalConfigLayer)
+        this.loadGrapherJson(mergedConfig)
+    }
+
+    async sendPatches(patches: VariableAnnotationPatch[]): Promise<void> {
+        await this.context.admin.rawRequest(
+            "/api/variable-annotations",
+            JSON.stringify(patches),
+            "PATCH"
+        )
+    }
+
+    /** Used to send the inverse of a patch to undo a change that has previously been performed
+     */
+    async sendReversedPatches(
+        patches: VariableAnnotationPatch[]
+    ): Promise<void> {
+        const reversedPatches = patches.map((patch) => ({
+            ...patch,
+            newValue: patch.oldValue,
+            oldValue: patch.newValue,
+        }))
+        await this.sendPatches(reversedPatches)
+    }
+
+    /** Performs an undo step - i.e. applies it to the richDataRows object,
```

Comment:
I find the naming as an UndoStep pretty confusing - it's an undoable step, but it's not being undone at this point.

Reply:
Fair enough - I renamed it to Action instead, is that better?